File: packages/endo-os/src/v8-host/microphone.cc
// Microphone (ALSA) Capability Binding
//
// Exposes an ALSA audio capture device as a streaming capability.
// Audio flows as Uint8Array chunks of PCM data.
//
// Like the camera, a microphone capability can be passed across
// chat — the recipient gets a live audio stream.  Revoke the
// capability reference and audio stops flowing.
//
// Uses the ALSA userspace API directly (no PulseAudio).
// On the minimal Endo OS kernel, ALSA is the only audio layer.
//
// JS API:
//   const mic = __openMicrophone('hw:0,0');
//   mic.sampleRate();             // → number (e.g., 44100)
//   mic.channels();               // → number (e.g., 1 or 2)
//   mic.bitsPerSample();          // → number (e.g., 16)
//   const chunk = mic.read(frames); // → Uint8Array (PCM data)
//   mic.start();
//   mic.stop();
//   mic.close();
//   mic.help();
//
// Note: This binding uses raw ALSA ioctls via /dev/snd/pcmC*D*c
// rather than libasound, keeping the binary dependency-free.
// For Phase 0, we provide a simplified implementation using
// the OSS-compatible /dev/dsp interface or direct ALSA.

#include "devices.h"

#include <cstdio>
#include <cstring>
#include <cerrno>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <linux/soundcard.h>

namespace endo {
namespace devices {

struct MicrophoneState {
  int fd;
  uint32_t sample_rate;
  uint32_t channels;
  uint32_t bits_per_sample;
  bool started;
};

static MicrophoneState* GetMicState(
    const v8::FunctionCallbackInfo<v8::Value>& info) {
  v8::Local<v8::External> wrap =
      v8::Local<v8::External>::Cast(info.This()->GetInternalField(0));
  return static_cast<MicrophoneState*>(wrap->Value());
}

static void MicSampleRate(
    const v8::FunctionCallbackInfo<v8::Value>& info) {
  info.GetReturnValue().Set(
      static_cast<double>(GetMicState(info)->sample_rate));
}

static void MicChannels(
    const v8::FunctionCallbackInfo<v8::Value>& info) {
  info.GetReturnValue().Set(
      static_cast<double>(GetMicState(info)->channels));
}

static void MicBitsPerSample(
    const v8::FunctionCallbackInfo<v8::Value>& info) {
  info.GetReturnValue().Set(
      static_cast<double>(GetMicState(info)->bits_per_sample));
}

// mic.read(frames) → Uint8Array (PCM data)
static void MicRead(const v8::FunctionCallbackInfo<v8::Value>& info) {
  v8::Isolate* isolate = info.GetIsolate();
  v8::HandleScope handle_scope(isolate);
  MicrophoneState* state = GetMicState(info);

  uint32_t frames = 1024;
  if (info.Length() >= 1 && info[0]->IsNumber()) {
    frames = info[0]->Uint32Value(isolate->GetCurrentContext())
                 .FromMaybe(1024);
  }

  // Calculate byte size: frames * channels * (bits/8).
  size_t frame_size =
      state->channels * (state->bits_per_sample / 8);
  size_t byte_count = frames * frame_size;

  if (byte_count > 4 * 1024 * 1024) {
    isolate->ThrowException(v8::Exception::RangeError(
        v8::String::NewFromUtf8Literal(isolate,
            "read size too large (max 4MB)")));
    return;
  }

  std::unique_ptr<v8::BackingStore> backing =
      v8::ArrayBuffer::NewBackingStore(isolate, byte_count);

  ssize_t n = read(state->fd, backing->Data(), byte_count);
  if (n < 0) {
    if (errno == EAGAIN) {
      v8::Local<v8::ArrayBuffer> buf =
          v8::ArrayBuffer::New(isolate, 0);
      info.GetReturnValue().Set(v8::Uint8Array::New(buf, 0, 0));
      return;
    }
    isolate->ThrowException(v8::Exception::Error(
        v8::String::NewFromUtf8Literal(isolate,
            "microphone read failed")));
    return;
  }

  v8::Local<v8::ArrayBuffer> ab =
      v8::ArrayBuffer::New(isolate, std::move(backing));
  info.GetReturnValue().Set(
      v8::Uint8Array::New(ab, 0, static_cast<size_t>(n)));
}

static void MicStart(const v8::FunctionCallbackInfo<v8::Value>& info) {
  MicrophoneState* state = GetMicState(info);
  // For OSS, reading implicitly starts capture.
  state->started = true;
  fprintf(stdout, "endo-init: Microphone capture started\n");
}

static void MicStop(const v8::FunctionCallbackInfo<v8::Value>& info) {
  MicrophoneState* state = GetMicState(info);
  // Reset the device to stop capture.
  ioctl(state->fd, SNDCTL_DSP_RESET, 0);
  state->started = false;
  fprintf(stdout, "endo-init: Microphone capture stopped\n");
}

static void MicClose(const v8::FunctionCallbackInfo<v8::Value>& info) {
  MicrophoneState* state = GetMicState(info);
  if (state->fd >= 0) {
    close(state->fd);
    state->fd = -1;
  }
}

static void MicHelp(const v8::FunctionCallbackInfo<v8::Value>& info) {
  info.GetReturnValue().Set(v8::String::NewFromUtf8Literal(
      info.GetIsolate(),
      "Microphone capability: sampleRate(), channels(), "
      "bitsPerSample(), read(frames), start(), stop(), close()"));
}

// __openMicrophone(device) → Microphone capability
//
// device: ALSA device string like 'hw:0,0' or OSS path like
//         '/dev/dsp'.  Defaults to '/dev/dsp' for simplicity.
static void OpenMicrophone(
    const v8::FunctionCallbackInfo<v8::Value>& info) {
  v8::Isolate* isolate = info.GetIsolate();
  v8::HandleScope handle_scope(isolate);

  const char* path = "/dev/dsp";
  v8::String::Utf8Value custom_path(isolate, info[0]);
  if (info.Length() >= 1 && info[0]->IsString()) {
    path = *custom_path;
  }

  int fd = open(path, O_RDONLY);
  if (fd < 0) {
    char msg[256];
    snprintf(msg, sizeof(msg), "Failed to open microphone: %s", path);
    isolate->ThrowException(v8::Exception::Error(
        v8::String::NewFromUtf8(isolate, msg,
                                v8::NewStringType::kNormal)
            .ToLocalChecked()));
    return;
  }

  // Configure audio format via OSS ioctls.
  int format = AFMT_S16_LE;
  ioctl(fd, SNDCTL_DSP_SETFMT, &format);

  int channels = 1; // mono
  ioctl(fd, SNDCTL_DSP_CHANNELS, &channels);

  int sample_rate = 44100;
  ioctl(fd, SNDCTL_DSP_SPEED, &sample_rate);

  auto* state = new MicrophoneState{
    fd,
    static_cast<uint32_t>(sample_rate),
    static_cast<uint32_t>(channels),
    16, // S16_LE = 16 bits
    false
  };

  fprintf(stdout,
      "endo-init: Microphone %s: %uHz %uch %ubit\n",
      path, state->sample_rate, state->channels,
      state->bits_per_sample);

  // Build capability object.
  v8::Local<v8::ObjectTemplate> tmpl = v8::ObjectTemplate::New(isolate);
  tmpl->SetInternalFieldCount(1);

  auto Set = [&](const char* name, v8::FunctionCallback cb) {
    tmpl->Set(v8::String::NewFromUtf8(isolate, name,
                                      v8::NewStringType::kNormal)
                  .ToLocalChecked(),
              v8::FunctionTemplate::New(isolate, cb));
  };

  Set("sampleRate", MicSampleRate);
  Set("channels", MicChannels);
  Set("bitsPerSample", MicBitsPerSample);
  Set("read", MicRead);
  Set("start", MicStart);
  Set("stop", MicStop);
  Set("close", MicClose);
  Set("help", MicHelp);

  v8::Local<v8::Object> obj =
      tmpl->NewInstance(isolate->GetCurrentContext()).ToLocalChecked();
  obj->SetInternalField(0, v8::External::New(isolate, state));

  info.GetReturnValue().Set(obj);
}

void InstallMicrophone(v8::Isolate* isolate,
                       v8::Local<v8::ObjectTemplate> global) {
  global->Set(
      v8::String::NewFromUtf8Literal(isolate, "__openMicrophone"),
      v8::FunctionTemplate::New(isolate, OpenMicrophone));
}

}  // namespace devices
}  // namespace endo

File: packages/endo-os/src/devices/microphone.rs
// Microphone (OSS/ALSA) Capability — Rust binding
//
// Exposes an audio capture device as deno_core ops.
// Uses the OSS-compatible /dev/dsp interface for simplicity
// (ALSA provides OSS emulation via snd-pcm-oss).
//
// Audio flows as Vec<u8> chunks of raw PCM data.  From JS,
// each read() returns a Uint8Array of interleaved samples.
//
// Ownership: the fd is held by MicrophoneResource and closed
// on drop.  No double-close, no use-after-free.

use std::cell::RefCell;
use std::fs::{File, OpenOptions};
use std::io::Read;
use std::os::unix::io::AsRawFd;

use anyhow::{anyhow, Result};
use deno_core::error::AnyError;
use deno_core::op2;
use deno_core::OpState;
use deno_core::Resource;
use deno_core::ResourceId;
use serde::Serialize;

pub struct MicrophoneResource {
    file: RefCell<File>,
    sample_rate: u32,
    channels: u32,
    bits_per_sample: u32,
}

impl Resource for MicrophoneResource {
    fn name(&self) -> std::borrow::Cow<str> {
        "microphone".into()
    }
}

// OSS ioctl numbers.
const SNDCTL_DSP_SETFMT: libc::c_ulong = 0xC0045005;
const SNDCTL_DSP_CHANNELS: libc::c_ulong = 0xC0045006;
const SNDCTL_DSP_SPEED: libc::c_ulong = 0xC0045002;
const SNDCTL_DSP_RESET: libc::c_ulong = 0x00005000;
const AFMT_S16_LE: i32 = 0x00000010;

#[derive(Serialize)]
pub struct MicInfo {
    pub sample_rate: u32,
    pub channels: u32,
    pub bits_per_sample: u32,
}

/// Open a microphone device.  Returns a resource ID.
#[op2(fast)]
pub fn op_open_microphone(
    state: &mut OpState,
    #[string] path: &str,
) -> Result<u32, AnyError> {
    let file = OpenOptions::new()
        .read(true)
        .open(path)
        .map_err(|e| anyhow!("Failed to open microphone {path}: {e}"))?;

    let fd = file.as_raw_fd();

    // Configure audio format via OSS ioctls.
    let mut format = AFMT_S16_LE;
    let mut channels: i32 = 1; // mono
    let mut sample_rate: i32 = 44100;

    unsafe {
        libc::ioctl(fd, SNDCTL_DSP_SETFMT, &mut format);
        libc::ioctl(fd, SNDCTL_DSP_CHANNELS, &mut channels);
        libc::ioctl(fd, SNDCTL_DSP_SPEED, &mut sample_rate);
    }

    println!(
        "endo-init: Microphone {path}: {sample_rate}Hz {channels}ch 16bit"
    );

    let rid = state.resource_table.add(MicrophoneResource {
        file: RefCell::new(file),
        sample_rate: sample_rate as u32,
        channels: channels as u32,
        bits_per_sample: 16,
    });

    Ok(rid)
}

/// Get microphone info.
#[op2]
#[serde]
pub fn op_mic_info(
    state: &mut OpState,
    #[smi] rid: ResourceId,
) -> Result<MicInfo, AnyError> {
    let resource = state.resource_table.get::<MicrophoneResource>(rid)?;
    Ok(MicInfo {
        sample_rate: resource.sample_rate,
        channels: resource.channels,
        bits_per_sample: resource.bits_per_sample,
    })
}

/// Read PCM audio frames.  Returns raw bytes (interleaved S16LE).
#[op2]
#[buffer]
pub fn op_mic_read(
    state: &mut OpState,
    #[smi] rid: ResourceId,
    #[smi] frames: u32,
) -> Result<Vec<u8>, AnyError> {
    let resource = state.resource_table.get::<MicrophoneResource>(rid)?;
    let mut file = resource.file.borrow_mut();

    let frame_size = resource.channels * (resource.bits_per_sample / 8);
    let byte_count = (frames as usize) * (frame_size as usize);
    let byte_count = byte_count.min(4 * 1024 * 1024); // 4 MB max

    let mut buf = vec![0u8; byte_count];
    let n = file.read(&mut buf)
        .map_err(|e| anyhow!("microphone read failed: {e}"))?;
    buf.truncate(n);

    Ok(buf)
}

/// Close the microphone.
#[op2(fast)]
pub fn op_mic_close(
    state: &mut OpState,
    #[smi] rid: ResourceId,
) -> Result<(), AnyError> {
    state.resource_table.close(rid)
        .map_err(|_| anyhow!("invalid microphone resource"))?;
    Ok(())
}
